### ISSUE TITLE
States are not visible in High Contrast views

### ISSUE THREAD

**Issue by karlgroves** [WordPress/gutenberg#15326]
States are not visible in High Contrast views

-   **Severity:** High
-   **Affected Populations:**
    -   Low-Vision
    -   Motor Impaired
    -   Cognitively Impaired
-   **Platform(s):**
    -   Windows - High Contrast
-   **Components affected:**
    -   Block Editing
    -   Block Panel
    -   Document Panel
    -   Publish and Unpublish

#### Issue description

Keyboard users who rely on Windows High Contrast themes cannot see their
focus or the pressed state on parts of the components listed.
Specifically, the focus states cannot be differentiated on the "AM"
and "PM" buttons, the next-month and previous-month buttons, or the
"Move to Trash" buttons, and there is no visual differentiation
between pressed and not on the "AM" and "PM" buttons.

All interactive elements should have visible and obvious state
indicators, so that sighted keyboard users have a clear indication of
which element has focus, and which state it's in.

##### Issue Code

```
    /* invisible pressed state of AM/PM buttons */


    ...,


    .components-datetime__time .components-datetime__time-wrapper .components-datetime__time-pm-button.is-toggled {


        ...


        box-shadow: inset 0 2px 5px -3px #555d66;


    }





    /* invisible focus states on Next/Previous Month buttons */


    /* default */


    .DayPickerNavigation_button__default {


        border: 1px solid #e4e7e7;


        ...


    }


    /* focussed */


    .DayPickerNavigation_button__default:focus, .DayPickerNavigation_button__default:hover {


        border: 1px solid #c4c4c4;


    }





    /* Move to Trash button */


    .components-button:focus:enabled {


        ...


       box-shadow: inset 0px 0px 0px 1px #6c7781, inset 0px 0px 0px 2px #fff;


        outline: 2px solid transparent;


        outline-offset: -2px;


    }





    /* buttons using SVG */


    <button type="button" aria-label="Heading 6" data-subscript="6" aria-pressed="true" class="...l is-active">


        <svg aria-hidden="true" role="img" focusable="false" ...>...</svg></button>


    ...


    .components-toolbar__control.components-button:not(:disabled).is-active>svg {


        outline: 0;


        color: #fff;


        box-shadow: none;


        background: #555d66;


    }





    /* color pickers */


    .components-color-palette__item.is-active {


        box-shadow: inset 0 0 0 4px;


    }
```

#### Remediation Guidance

Darken the `border` or `box-shadow` colors to meet minimum color
ratios for non-text contrast.

Add a thick transparent outline to focussed buttons for Windows High
Contrast users. For buttons such as "Move to Trash" where there is an
`outline`, remove the negative `outline-offset` which is currently
preventing the new outline from being visible against the border.

For the toggled or `pressed` state of the AM/PM buttons, use a visible
difference in high contrast to show the `pressed` state (such as a
persistent outline) and something different such as a dotted outline to
reveal the focussed state.

The Recommended Code suggests one way of achieving this, but the
important take-away is that changes in border color are not revealed in
High Contrast views, whereas changes in border thickness and styles are
visible.

##### Recommended Code

```
    /* pressed state of AM/PM buttons */


    ...,


    .components-datetime__time .components-datetime__time-wrapper .components-datetime__time-pm-button.is-toggled {


        ...


        box-shadow: inset 0 2px 5px -3px #00435d;


        outline: 3px solid transparent; /* or a thicker border than the non-toggled button... */


    }


    .components-datetime__time .components-datetime__time-wrapper .components-datetime__time-pm-button.is-toggled:focus {


        ...


        outline: 1px dotted transparent;


        outline-offset: 3px;


    }





    /* invisible focus states on Next/Previous Month buttons */


    /* default */


    .DayPickerNavigation_button__default {


        border: 1px solid #e4e7e7;


        ...


    }


    /* focussed */


    .DayPickerNavigation_button__default:focus, .DayPickerNavigation_button__default:hover {


        border: 1px solid #8E8E8E;


        outline: 3px solid transparent;


    }


    /* Move to Trash button */


    .components-button:focus:enabled {


        box-shadow: inset 0px 0px 0px 1px #00435d, inset 0px 0px 0px 2px #fff;


        outline: 2px solid transparent;


    }





    /* buttons using SVG */


    .components-toolbar__control.components-button:not(:disabled).is-active>svg {


        outline: 0;


        color: #fff;


        box-shadow: none;


        background: #555d66;


        border: 1px solid transparent;


    }








    /* color pickers */


    .components-color-palette__item.is-active {


        box-shadow: inset 0 0 0 4px;


        border: 1px dotted transparent;


    }





    <!-- displayed color -->


    <span class="components-base-control__label">Background Color


        <span class="component-color-indicator" aria-label="(current background color: Vivid red)" style="color: transparent;border: 0;background: rgb(207, 46, 46);">Vivid red</span>


    </span>

```

#### Relevant standards

-   [1.3.3 Sensory Characteristics
    (Level A)](https://www.w3.org/TR/WCAG20/#content-structure-separation-understanding)
-   [2.4.7 Focus Visible (Level
    AA)](https://www.w3.org/TR/WCAG20/#navigation-mechanisms-focus-visible)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-29 in Tenon's report

**Comment by enriquesanchez** [comment]
This was [discussed during design triage](https://wordpress.slack.com/archives/C02S78ZAL/p1556822844486800). 

We'll investigate further and try to capture some screenshots for reference.

**Comment by melchoyce** [comment]
Screenshots from the full report:

![image](https://user-images.githubusercontent.com/2846578/57184330-ff49df00-6e6e-11e9-82d9-43b411a75da4.png)
![image](https://user-images.githubusercontent.com/2846578/57184332-0cff6480-6e6f-11e9-986b-152a32833ee6.png)

**Comment by enriquesanchez** [comment]
More screenshots from the report:

<img width="699" alt="Screen Shot 2019-05-06 at 12 57 46 PM" src="https://user-images.githubusercontent.com/3276087/57244710-9abd8a00-6ffe-11e9-9c53-811140cd13b0.png">
![Uploading Screen Shot 2019-05-06 at 12.50.44 PM.png…]()

**Comment by enriquesanchez** [comment]
Here are some side by side comparisons of the Media Library between regular and high contrast mode in Windows 10.

1. Nothing focused:
<img width="1739" alt="Screen Shot 2019-05-06 at 1 59 36 PM" src="https://user-images.githubusercontent.com/3276087/57248287-99dd2600-7007-11e9-91ee-838716432c23.png">

2. Focus on 'Media Library' tab:
<img width="1739" alt="Screen Shot 2019-05-06 at 2 00 43 PM" src="https://user-images.githubusercontent.com/3276087/57248335-b8dbb800-7007-11e9-8599-5dc78a8bfac3.png">

3. Focus on 'All dates' dropdown:
<img width="1741" alt="Screen Shot 2019-05-06 at 2 00 55 PM" src="https://user-images.githubusercontent.com/3276087/57248355-c729d400-7007-11e9-8a52-76196387b55e.png">

4. Focus on the first image of the list:
<img width="1739" alt="Screen Shot 2019-05-06 at 2 01 07 PM" src="https://user-images.githubusercontent.com/3276087/57248391-dc9efe00-7007-11e9-878f-ee636249d09e.png">

**Comment by afercia** [comment]
Worth noting that a first part of CSS improvements for Windows High Contrast mode landed in core in https://core.trac.wordpress.org/changeset/44544 see the related ticket https://core.trac.wordpress.org/ticket/41286

Further improvements are tracked in a new ticket: https://core.trac.wordpress.org/ticket/45910

I'm pretty sure there are some "custom" elements in the media views that have no visible focus in High Contrast Mode. This part should be tested and moved to Trac. The part related to High Contrast Mode in Gutenberg should stay.

**Comment by tellthemachines** [comment]
The Gutenberg-specific High Contrast Mode issues have been fixed with #16554 ; closing this ticket.
